src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1991,6 +1991,44 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
             return new PointerTypeSymbol(TypeWithAnnotations.Create(elementType, elementNullableAnnotation));
         }
 
+        internal FunctionPointerTypeSymbol CreateFunctionPointerTypeSymbol(
+            TypeWithAnnotations returnType,
+            ImmutableArray<TypeWithAnnotations> parameterTypes,
+            RefKind returnRefKind,
+            ImmutableArray<RefKind> parameterRefKinds)
+        {
+            if (returnType.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(returnType));
+            }
+
+            if (parameterTypes.IsDefault || parameterTypes.Any(param => param.IsDefault))
+            {
+                throw new ArgumentNullException(nameof(parameterTypes));
+            }
+
+            if (parameterRefKinds.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(parameterRefKinds));
+            }
+
+            if (parameterRefKinds.Length != parameterTypes.Length)
+            {
+                // Given {0} parameter types and {1} parameter ref kinds. These must be the same.
+                throw new ArgumentException(string.Format(CSharpResources.NotSameNumberParameterTypesAndRefKinds, parameterTypes.Length, parameterRefKinds.Length));
+            }
+
+            if (returnRefKind == RefKind.Out)
```

Comment:
I'm unsure how much validation we should be doing here. For example, should we be validating that `void` is not paired with a refkind? Or that `void` is not provided as a parameter type?